### DISCUSSION TITLE
Changing the dGraphicsAlphaBits value via the MagickWand API

### DISCUSSION THREAD

**Discussion by jonn** [ImageMagick/ImageMagick#6557] [Help] [answered]
### ImageMagick version

7.1.1-15

### Operating system, version and so on

macOS Monterey 

### Description

Hello ImageMagick community,

I'm working on an app that will validate a 3rd party provided PDF for print purposes and create a print ready TIFF from the the PDF. The app is written in Go and I'm using the MagickWand API. Of the ten PDF example files I was given, ImageMagick has produced exactly what I needed and hoped for (kudos ImageMagick developers! You've created something incredible), in all cases but one. The one exception is an output file that displays semi-transparency not visibly present in the source PDF.

Screenshot of Preview app displaying the original PDF:
<img width="280" alt="screenshot-source" src="https://github.com/ImageMagick/ImageMagick/assets/106102/d007e82c-a4c0-4b28-9f9d-304b216ee96e">

Screenshot of Preview app displaying the IM produced TIFF:
<img width="280" alt="screenshot-im-output" src="https://github.com/ImageMagick/ImageMagick/assets/106102/92ba425d-f237-49fa-9781-e93ff9e9d23c">

To confirm this was not a result of something I had done incorrectly with the MagickWand API, I switched to using the ImageMagick `convert` binary. This command produced the same results:

`convert -verbose -depth 8 -density 300 -units PixelsPerInch source.pdf -compress LZW output.tif`

...along with the following output:

```
/var/folders/5j/75f2l8d50_jgps5mwrkq3j2h0000gn/T/magick-xSgOF-E6h4cpKkCFHQ6wM5Is_HuUWK5f1 PAM 2788x3538 2788x3538+0+0 8-bit ColorSeparation CMYK 37.6281MiB 1.220u 0:00.211
source.pdf AI 2788x3538 2788x3538+0+0 8-bit ColorSeparation CMYK 37.6281MiB 0.060u 0:00.008
[ghostscript library 10.01] -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pamcmyk32' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r300x300' -dPrinted=false  '-sOutputFile=/var/folders/5j/75f2l8d50_jgps5mwrkq3j2h0000gn/T/magick-xSgOF-E6h4cpKkCFHQ6wM5Is_HuUWK5f%d' '-f/var/folders/5j/75f2l8d50_jgps5mwrkq3j2h0000gn/T/magick-1J7-tWMLkfIrftiszh7EG1hIruNWFVte' '-f/var/folders/5j/75f2l8d50_jgps5mwrkq3j2h0000gn/T/magick-rzEZB3dSNV40usVB60V2mUaFBBnUptiT'source.pdf=>output-3-nocrop.tif AI 2788x3538 2788x3538+0+0 8-bit CMYK 3.15009MiB 1.790u 0:00.351
```

I then moved to using GhostScript directly and determine the problem is in the -dGraphicsAlphaBits=4 option. If I remove that option from the `gs` command, or change that option value to 2, the output image matches what I see in the display of the original PDF.

Also, I tested export of the PDF from both Photoshop and Preview and did not see this problem.

Question 1) is this an ImageMagick bug? It seems like a bug as the visual change is not present when using Photoshop, Preview, or GhostScript without setting the -dGraphicsAlphaBits option to 4. It seems like IM is getting this option wrong. But I'm new to using IM, so I could be misunderstanding something. I appreciate guidance from the community before I file a bug.

Question 2) is there some way I can set, change, or remove the -dGraphicsAlphaBits option value via the MagickWand API? If I can remove it from the generated `gs` command, or set it to 2, it seems the problem will go away.

Thank you for any help or direction you can provide!

**Comment by fmw42** [comment]
Ghostscript does not handle CMYK very well. Convert to sRGB by adding -colorspace sRGB before reading the input, i.e. right after magick and before the PDF file.

**Comment by jonn** [comment]
Ah, thank you for the insight @fmw42. These files are created for print. The PDFs are provide in CMYK and the produced TIFFs need to be CMYK. If Ghostscript does not handle CMYK well, does that mean it, and ImageMagick, are the wrong tools for the job? 

Thanks again.

**Reply by fmw42** [reply]
Convert the PDF to srgb, then read it with Ghostscript. Then convert that from srgb back to CMYK with profiles.

**Reply by fmw42** [reply]
Post and original PDF, not a screen snap, so we can see what it looks like and test commands.

**Reply by jonn** [reply]
Thank you for being willing to look at the original file. It's not mine to share publicly. Let me see if I can clear it with the owner tomorrow. Much appreciation!

**Reply by fmw42** [reply]
Get a PDF that is CMYK which is not proprietary or make one.

**Comment by jonn** [comment]
Thank you for your help and your patience. Attached is a PDF that exhibits the problem when run through ImageMagick.

[source_cropped.pdf](https://github.com/ImageMagick/ImageMagick/files/12347042/source_cropped.pdf)

The following ImageMagick `convert` command exhibits the unexpected output:

`convert -depth 8 -density 300 -units PixelsPerInch source_cropped.pdf -compress LZW output_test.2A.tif`

The following Ghostscript command produces a TIFF with the expected appearance:

`gs -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=tiff24nc' -dTextAlphaBits=4 '-r300x300' -dPrinted=false '-sOutputFile=output_test.2B.tif' -sCompression=lzw source_cropped.pdf -c quit -q`

**Comment by fmw42** [comment, marked as answer]
On IM 7, use magick not convert.

When I run this command on IM 7.1.0.15 with GS 9.56.1. (note that I moved -depth 8 until after reading the PDF)

magick -density 300 -units PixelsPerInch source_cropped.pdf -depth 8 -compress LZW output_test.2A.tif

I get

[output_test.2A.tif.zip](https://github.com/ImageMagick/ImageMagick/files/12347100/output_test.2A.tif.zip)


Is this not correct?  What version of Ghostscript are you using?


If that gs command works for you, you can edit your delegates.xml file so that command is being use.  I am not an expert on how to edit it. Perhaps one of the other users can inform you what to do.

**Reply by jonn** [reply]
That is correct! Your output file looks like what I expect.

My version of gs is 10.01.2. When I run the same `magick` command you just posted, I get the same unexpected output I get with the `covert` command.

Let me see what happens when I roll back the gs version.

**Reply by fmw42** [reply]
I have heard of issues with GS 10.  But I know little about running it standalone.  If that works, you should be able to edit the delegates.xml file to use those parameters.

**Reply by dpthurst** [reply]
Editing delegates.xml won't work, since it expects an integer parameter that gets passed in. The logic in question is in `coders/pdf.c` (line 653):

> ``  (void) FormatLocaleString(command,MagickPathExtent,
>     GetDelegateCommands(delegate_info),
>     read_info->antialias != MagickFalse ? 4 : 1,
>     read_info->antialias != MagickFalse ? 4 : 1,density,options,filename,
>     postscript_filename,input_filename);``

If you change the code to unconditionally pass 1 to the two arguments with the logic it should work.

**Reply by jonn** [reply]
Turns out version locking Ghostscript via brew isn't as easy as I expected. Still working on that as possible solution.

I a have also been unsuccessful editing delegates.xml. I not sure I know what I'm doing there and need to read-up, but it seems I am running up against what @dpthurst is talking about: if I remove the -dGraphicsAlphaBits option, or try to set it to 2, I get a `zsh: abort`.

**Reply by jonn** [reply]
Following up on this... after giving up on a delegates.xml solution, and trying a PDF handling case with direct Ghostscript usage via a go bindings-for-Ghostscript package, only to discover that it did not support gs 10... I  ended up building and deploying the app in a container where I was able to build imagemagick 7.1.1-16 from source with the distro provided libgs of version 9.56.1. With this combo, the PDF semi-transparent layering problem went away. Additionally, another problem I was seeing, with getting incorrect dimension/resolution numbers on certain PDFs, also went away. I'm sticking with Ghostscript 9 for now.

Thank you @fmw42 and @dpthurst for the help on this.

**Reply by fmw42** [reply]
I suggest you post a bug report for GS 10 to the GS developers.

**Comment by dpthurst** [comment]
I've encountered a similar bug. You should be able to fix it by turning off anti-aliasing ('+antialias' on the command line), though it's possible that will cause other problems.

On the Ghostscript bug tracker, the developers give advice to not use either GraphicsAlphaBits or TextAlphaBits with PDF files; see (eg) [here](https://bugs.ghostscript.com/show_bug.cgi?id=697860).

**Reply by jonn** [reply]
Thank you! Turning off anti-aliasing does make the semi-transparent layering problem go away. But yes, turning off anti-aliasing results in suboptimal text in the TIFF output.

And thank you for the bug pointer. I'm inching toward clarity on what is happening and possible solutions/work-arounds.
